common/flatpak-dir.c

```diff
@@ -1899,6 +1909,405 @@ get_system_locations (GCancellable *cancellable,
+  config->branch = g_strdup ("master");
+  config->install = TRUE;
+
+  return config;
+}
+
+static void
+preinstall_config_free (PreinstallConfig *config)
+{
+  g_clear_pointer (&config->name, g_free);
+  g_clear_pointer (&config->branch, g_free);
+  g_clear_pointer (&config->collection_id, g_free);
+  g_free (config);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (PreinstallConfig, preinstall_config_free)
+
+static void
+flatpak_preinstall_config_free (FlatpakPreinstallConfig *preinstall)
+{
+  g_clear_pointer (&preinstall->ref, flatpak_decomposed_unref);
+  g_clear_pointer (&preinstall->collection_id, g_free);
+  g_free (preinstall);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (FlatpakPreinstallConfig, flatpak_preinstall_config_free)
+
+static void
+flatpak_parse_preinstall_config_file (GKeyFile   *keyfile,
+                                      GHashTable *configs)
+{
+  g_auto(GStrv) groups = NULL;
+
+  groups = g_key_file_get_groups (keyfile, NULL);
+
+  for (int i = 0; groups[i] != NULL; i++)
+    {
+      const char *group_name = groups[i];
+      const char *name;
+      g_autoptr(PreinstallConfig) config = NULL;
+      g_autoptr(GError) local_error = NULL;
+      g_autofree char *owned_name = NULL;
+      g_autofree char *branch = NULL;
+      gboolean is_runtime = FALSE;
+      g_autofree char *collection_id = NULL;
+      gboolean install = TRUE;
+
+      if (!g_str_has_prefix (group_name, FLATPAK_PREINSTALL_GROUP_PREFIX) ||
+          *(group_name + strlen (FLATPAK_PREINSTALL_GROUP_PREFIX)) == '\0')
+        {
+          g_info ("Skipping unknown group %s", group_name);
+          continue;
+        }
+
+      name = group_name + strlen (FLATPAK_PREINSTALL_GROUP_PREFIX);
+
+      if (!g_hash_table_steal_extended (configs, name,
+                                        (gpointer *)&owned_name,
+                                        (gpointer *)&config))
+        {
+          config = preinstall_config_new (name);
+          owned_name = g_strdup (name);
+        }
+
+      branch = g_key_file_get_string (keyfile,
+                                      group_name,
+                                      FLATPAK_PREINSTALL_BRANCH_KEY,
+                                      NULL);
+      if (branch)
+        {
+          g_clear_pointer (&config->branch, g_free);
+          if (*branch != '\0')
+            config->branch = g_steal_pointer (&branch);
+        }
+
+      is_runtime = g_key_file_get_boolean (keyfile,
+                                           group_name,
+                                           FLATPAK_PREINSTALL_IS_RUNTIME_KEY,
+                                           &local_error);
+      if (!local_error)
+        {
+          config->is_runtime = is_runtime;
+        }
+      else if (!g_error_matches (local_error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_KEY_NOT_FOUND))
+        {
+          g_info ("Invalid file format, %s is not a boolean",
+                  FLATPAK_PREINSTALL_IS_RUNTIME_KEY);
+          continue;
```

Comment:
I don't like how we modify the config and then fail halfway in and ignore the rest. I think we should parse all the options and then apply the in the end as a unit.

Reply:
We only add it to the hashtable if all parsing succeeded, but that also means we would drop a config if a higher priority config fails to parse an option, which isn't great. Changed this to never `continue` and thus never fail halfway.